File: apps/api/internal/shooter/waves.go
package shooter

import "slices"

func (sim *simulation) spawnWave() {
	for _, spawn := range sim.config.Wave.Spawns {
		count := max(1, spawn.Count)
		every := max(1, spawn.IntervalTicks)
		for occurrence := 0; occurrence < count; occurrence++ {
			if sim.tick-1 != spawn.AtTick+occurrence*every {
				continue
			}
			sim.spawnEnemy(spawn.EnemyID, formationX(spawn.Formation, occurrence, count, sim.tick))
		}
	}
	sim.spawnLatePressure()
}

// spawnLatePressure keeps a survival segment active after its authored opening
// formations have finished. The selected chassis and formation are derived
// only from the segment seed and cycle number, so Go replay and browser
// prediction can reproduce the same late-show pressure without more content
// payload. Boss segments never receive these reinforcements.
func (sim *simulation) spawnLatePressure() {
	if sim.config.Boss != nil || len(sim.config.Wave.Spawns) == 0 || len(sim.enemies) >= sim.config.Limits.Enemies {
		return
	}
	if len(sim.enemies) == 0 && len(sim.enemyProjectiles) == 0 {
		sim.pressureQuietTicks++
	} else {
		sim.pressureQuietTicks = 0
	}
	lastAuthoredTick := 0
	pool := make([]string, 0, len(sim.config.Wave.Spawns))
	for _, spawn := range sim.config.Wave.Spawns {
		lastAuthoredTick = max(lastAuthoredTick, spawn.AtTick+(max(1, spawn.Count)-1)*max(1, spawn.IntervalTicks))
		if !slices.Contains(pool, spawn.EnemyID) {
			pool = append(pool, spawn.EnemyID)
		}
	}
	start := lastAuthoredTick + 90
	interval := 120
	if sim.config.EncoreLevel >= 1 {
		interval = 105
	}
	elapsed := sim.tick - 1 - start
	finalTick := sim.config.DurationTicks - 90
	regularPulse := elapsed >= 0 && elapsed%interval == 0 && sim.tick-1 <= sim.config.DurationTicks-60
	lastRegular := start
	if finalTick > start {
		lastRegular = start + (finalTick-start)/interval*interval
	}
	finalPulse := sim.tick-1 == finalTick && finalTick != lastRegular
	emergencyPulse := sim.tick-1 > 30 && sim.pressureQuietTicks >= 90
	if !regularPulse && !finalPulse && !emergencyPulse {
		return
	}
	cycle := max(0, elapsed/interval)
	seedOffset := int(seedFromString(sim.config.Seed+":late-pressure") % uint32(len(pool)))
	enemyID := pool[(seedOffset+cycle)%len(pool)]
	formations := [...]string{"pincer", "sweep", "staggered", "fan"}
	formation := formations[(seedOffset+cycle)%len(formations)]
	count := 1
	if sim.config.EncoreLevel >= 1 && cycle&1 != 0 {
		count = 2
	}
	for index := 0; index < count && len(sim.enemies) < sim.config.Limits.Enemies; index++ {
		sim.spawnEnemy(enemyID, formationX(formation, index, count, sim.tick))
	}
	sim.pressureQuietTicks = 0
}

func formationX(formation string, index, count, tick int) int {
	center := ArenaWidth / 2
	step := 520
	switch formation {
	case "line":
		return center + (index*2-(count-1))*step/2
	case "fan":
		return center + (index*2-(count-1))*360/2
	case "staggered":
		return 520 + (index*760)%2560
	case "pincer":
		if index&1 == 0 {
			return 400 + (index/2)*220
		}
		return ArenaWidth - 400 - (index/2)*220
	case "center":
		return center + (index*2-(count-1))*180/2
	case "sweep":
		return 320 + ((tick*17 + index*540) % (ArenaWidth - 640))
	default:
		return center
	}
}

func (sim *simulation) spawnEnemy(specID string, authoredX int) {
	if len(sim.enemies) >= sim.config.Limits.Enemies {
		return
	}
	specIndex := -1
	for index, spec := range sim.config.Enemies {
		if spec.ID == specID {
			specIndex = index
			break
		}
	}
	if specIndex < 0 {
		return
	}
	x := authoredX
	if x <= 0 || x >= ArenaWidth {
		x = 320 + sim.random.intn(ArenaWidth-640)
	}
	x = clamp(x, enemyRadius, ArenaWidth-enemyRadius)
	spec := sim.config.Enemies[specIndex]
	// Encore changes formations and attack semantics; it deliberately does not
	// turn enemies into larger health bars.
	health := spec.Health
	sim.nextEnemyID++
	sim.enemies = append(sim.enemies, enemyEntity{id: sim.nextEnemyID, specIndex: specIndex, x: x, y: 500, health: health, maxHealth: health})
}

File: apps/miniapp/src/features/shooter/waves.ts
import {
  ENEMY_RADIUS,
  SHOOTER_WIDTH,
  clamp,
  goDivide,
} from "@/features/shooter/constants";
import { shooterSeedFromString } from "@/features/shooter/random";
import type { ShooterMutableState } from "@/features/shooter/types";

const formationX = (
  formation: string,
  index: number,
  count: number,
  tick: number,
): number => {
  const center = SHOOTER_WIDTH / 2;
  if (formation === "line") return center + ((index * 2 - (count - 1)) * 520) / 2;
  if (formation === "fan") return center + ((index * 2 - (count - 1)) * 360) / 2;
  if (formation === "staggered") return 520 + (index * 760) % 2_560;
  if (formation === "pincer") return index & 1 ? SHOOTER_WIDTH - 400 - Math.trunc(index / 2) * 220 : 400 + Math.trunc(index / 2) * 220;
  if (formation === "center") return center + ((index * 2 - (count - 1)) * 180) / 2;
  if (formation === "sweep") return 320 + (tick * 17 + index * 540) % (SHOOTER_WIDTH - 640);
  return center;
};

const spawnEnemy = (
  state: ShooterMutableState,
  specID: string,
  authoredX: number,
): void => {
  if (state.enemies.length >= state.config.limits.enemies) return;
  const specIndex = state.config.enemies.findIndex((spec) => spec.id === specID);
  if (specIndex < 0) return;
  let x = authoredX;
  if (x <= 0 || x >= SHOOTER_WIDTH) x = 320 + state.random.integer(SHOOTER_WIDTH - 640);
  x = clamp(x, ENEMY_RADIUS, SHOOTER_WIDTH - ENEMY_RADIUS);
  const spec = state.config.enemies[specIndex]!;
  const health = spec.health;
  state.nextEnemyID += 1;
  state.enemies.push({
    id: state.nextEnemyID,
    specIndex,
    x,
    y: 500,
    health,
    maxHealth: health,
    fireClock: 0,
    age: 0,
    phase: 0,
    warning: 0,
    volley: 0,
    marks: 0,
    boss: false,
  });
};

const spawnLatePressure = (state: ShooterMutableState): void => {
  if (
    state.config.boss ||
    state.config.wave.spawns.length === 0 ||
    state.enemies.length >= state.config.limits.enemies
  ) {
    return;
  }
  if (state.enemies.length === 0 && state.enemyProjectiles.length === 0) {
    state.pressureQuietTicks += 1;
  } else {
    state.pressureQuietTicks = 0;
  }
  let lastAuthoredTick = 0;
  const pool: string[] = [];
  for (const spawn of state.config.wave.spawns) {
    lastAuthoredTick = Math.max(
      lastAuthoredTick,
      spawn.at_tick +
        (Math.max(1, spawn.count) - 1) * Math.max(1, spawn.interval_ticks),
    );
    if (!pool.includes(spawn.enemy_id)) pool.push(spawn.enemy_id);
  }
  const start = lastAuthoredTick + 90;
  const interval = state.config.encore_level >= 1 ? 105 : 120;
  const elapsed = state.tick - 1 - start;
  const finalTick = state.config.duration_ticks - 90;
  const regularPulse =
    elapsed >= 0 &&
    elapsed % interval === 0 &&
    state.tick - 1 <= state.config.duration_ticks - 60;
  let lastRegular = start;
  if (finalTick > start) {
    lastRegular = start + goDivide(finalTick - start, interval) * interval;
  }
  const finalPulse = state.tick - 1 === finalTick && finalTick !== lastRegular;
  const emergencyPulse = state.tick - 1 > 30 && state.pressureQuietTicks >= 90;
  if (!regularPulse && !finalPulse && !emergencyPulse) return;

  const cycle = Math.max(0, goDivide(elapsed, interval));
  const seedOffset =
    shooterSeedFromString(`${state.config.seed}:late-pressure`) % pool.length;
  const enemyID = pool[(seedOffset + cycle) % pool.length]!;
  const formations = ["pincer", "sweep", "staggered", "fan"] as const;
  const formation = formations[(seedOffset + cycle) % formations.length]!;
  const count =
    state.config.encore_level >= 1 && (cycle & 1) !== 0 ? 2 : 1;
  for (
    let index = 0;
    index < count && state.enemies.length < state.config.limits.enemies;
    index += 1
  ) {
    spawnEnemy(
      state,
      enemyID,
      formationX(formation, index, count, state.tick),
    );
  }
  state.pressureQuietTicks = 0;
};

export const spawnWave = (state: ShooterMutableState): void => {
  for (const spawn of state.config.wave.spawns) {
    const count = Math.max(1, spawn.count);
    const every = Math.max(1, spawn.interval_ticks);
    for (let occurrence = 0; occurrence < count; occurrence += 1) {
      if (state.tick - 1 === spawn.at_tick + occurrence * every) {
        spawnEnemy(state, spawn.enemy_id, formationX(spawn.formation, occurrence, count, state.tick));
      }
    }
  }
  spawnLatePressure(state);
};
